include/sst/effects/Bonsai.h

```diff
@@ -0,0 +1,1202 @@
+                                               const float coef200, const float coef500,
+                                               float boost, float dist, float *__restrict srcL,
+                                               float *__restrict srcR, float *__restrict dstL,
+                                               float *__restrict dstR)
+{
+    float bufA alignas(16)[FXConfig::blockSize] = {};
+    float bufB alignas(16)[FXConfig::blockSize] = {};
+    float reused alignas(16)[FXConfig::blockSize] = {};
+    float branch1 alignas(16)[FXConfig::blockSize] = {};
+    float branch2 alignas(16)[FXConfig::blockSize] = {};
+    float branch3 alignas(16)[FXConfig::blockSize] = {};
+
+    const float dist01 = dist * 0.3333333333333333333333333;
+    const float distsq = dist * dist; // this will extend past 0-1
+    const float distinvsq = invsq(dist01);
+
+    float lerp1_block alignas(16)[FXConfig::blockSize] = {};
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 0], coef20,
+                                          rerange01(distinvsq, 0.025, 0.01), lerp1_block);
+    float lerp2_block alignas(16)[FXConfig::blockSize] = {};
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 1], coef20, rerange01(distsq, 0.125, 1.f),
+                                          lerp2_block);
+    float lerp3_block alignas(16)[FXConfig::blockSize] = {};
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 2], coef20, rerange01(distsq, 0.5, 10.f),
+                                          lerp3_block);
+    float lerp4_block alignas(16)[FXConfig::blockSize] = {};
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 3], coef20, rerange01(distsq, 1.f, 5.f),
+                                          lerp4_block);
+    float lerp5_block alignas(16)[FXConfig::blockSize] = {};
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 4], coef20,
+                                          rerange01(dist01, 0.075, 0.025), lerp5_block);
+    float lerp6_block alignas(16)[FXConfig::blockSize] = {};
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 5], coef20, rerange01(dist01, 0.1, 0.075),
+                                          lerp6_block);
+    float boost_block alignas(16)[FXConfig::blockSize] = {};
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 6], coef20, boost, boost_block);
+
+    onepole_hp_block<FXConfig::blockSize>(last[lastmin + 7], coef20, srcL, bufA);
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 8], coef50, bufA, bufB);
+    clip_tanh78_block<FXConfig::blockSize>(lerp1_block, bufB, bufB);
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 9], coef50, bufB, bufA);
+    mul_block<FXConfig::blockSize>(bufA, 20.f, branch1);
+    onepole_hp_block<FXConfig::blockSize>(last[lastmin + 10], coef30, srcL, bufA);
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 11], coef200, bufA, bufB);
+    onepole_lp_block<FXConfig::blockSize>(last[lastmin + 12], coef200, bufB, reused);
```

Comment:
Now I read this code, I really wonder if we want a little 'with last wrapper' object. That is something which houses the last value and calls down to a function.

I'm thinking like

```
template<(ugly arg for F)>
struct lastLift
{
   float L{0};
   template<ugly arg for vararg>
   inline void process(...) { F(L, ...); }
}
```

then we can wrap each of these in a last managing thing and not have all this index fragility on the last array.

Also I'm curious if this is fast enough to run realtime!

And also, wow, how are we ever going to debug or modify this. I'm not sure the structure is obvious from this code. Wonder if there's something we can do about that.

Let me ponder a little a cleverer way to write this perhaps.

Reply:
cleverer way to structure that is.

Reply:
yeah i should transcribe the reaktor core cell into something, since i had to refer to that a lot when writing this

Reply:
and yeah some way to say the next last index, and handle how much a function uses. someone suggested giving each function it's own last array, maybe i should change it to that